wodin_deploy/wodin_constellation.py

```diff
@@ -0,0 +1,117 @@
+        redis = redis_container(cfg)
+        odin_api = odin_api_container(cfg)
+        sites = sites_containers(cfg)
+        proxy = wodin_proxy_container(cfg)
+
+        containers = [redis, odin_api, *sites, proxy]
+
+        self.obj = Constellation(
+            "wodin",
+            cfg.container_prefix,
+            containers,
+            cfg.network,
+            {"redis-data": "redis-data", "wodin-config": "wodin-config"},
+            data=cfg,
+        )
+
+    def start(self, **kwargs):
+        self.obj.start(**kwargs)
+
+    def stop(self, **kwargs):
+        self.obj.stop(**kwargs)
+
+    def status(self):
+        self.obj.status()
+
+
+def redis_container(cfg):
+    redis = cfg.redis
+    redis_mounts = [ConstellationMount("redis-data", "/data")]
+    return ConstellationContainer(redis["name"], redis["ref"], mounts=redis_mounts, network=cfg.network)
+
+
+def odin_api_container(cfg):
+    odin_api = cfg.odin_api
+    return ConstellationContainer("api", odin_api["ref"], network=cfg.network)
+
+
+def get_wodin_container(cfg, site, site_dict):
+    wodin = cfg.wodin
+    prefix = cfg.container_prefix
+    wodin_mounts = [ConstellationMount("wodin-config", "/wodin/config")]
+    container_name = f"{wodin['name']}-{site}"
+
+    return ConstellationContainer(
+        container_name,
+        wodin["ref"],
+        mounts=wodin_mounts,
+        network=cfg.network,
+        entrypoint=[
+            "/wodin/docker/pull-site-and-start.sh",
+            site_dict["ref"],
+            site_dict["url"],
+            f"/wodin/config/{site}",
+            f"--redis-url=redis://{prefix}-{cfg.redis['name']}:6379",
+            f"--odin-api=http://{prefix}-api:8001",
+            f"--base-url=http://localhost/{site_dict['urlPath']}",
+            f"/wodin/config/{site}",
+        ],
+    )
+
+
+def sites_containers(cfg):
+    sites = cfg.sites
+    wodin_containers = []
+    for site in sites.keys():
+        site_dict = sites[site]
+        wodin_containers.append(get_wodin_container(cfg, site, site_dict))
+    return wodin_containers
+
+
+def create_index_page(sites):
+    html_strings = [
+        "<html>\n<head>\n<title>Wodin</title>\n</head>\n<body>\n",
+        "<h1>Wodin</h1>\n",
+        f"<p>There are {len(sites)} deployed configurations:</p>\n",
+        "<ul>\n",
+    ]
+    for site in sites.keys():
+        site_obj = sites[site]
+        html_strings.append(
+            "<li><a href='/{}'><code>{}</code></a>: {}</li>\n".format(
+                site_obj["urlPath"], site, site_obj["description"]
+            )
+        )
+    html_strings.append("</ul>\n</body>\n</html>")
+    return "".join(html_strings)
+
+
+def proxy_configure(container, cfg):
+    docker_util.exec_safely(container, "mkdir /wodin/root")
+    html_string = create_index_page(cfg.sites)
+    docker_util.string_into_container(html_string, container, "/wodin/root/index.html")
+    docker_util.exec_safely(container, "chmod +r /wodin/root/index.html")
```

Comment:
nginx cannot read a file that we create using `string_into_container` without this, any ideas on how to improve this? I wasnt to sure how else to add read permissions

Reply:
So looks like that is just a result of `string_into_container` it creates a file which only has read and write permission for the user who creates it. And no permissions for anyone else. So I think you'll have to chmod it. Elsewhere done this with an array like `docker_util.exec_safely(container, ["chmod", "+r", path])`.

Might be worth putting that path into a variable too and using that anywhere you reference it.

The alternative to chmod would be if you have the `index.html` file as a template on disk you can set the read permission on that and then copy it into the container.

Reply:
ahh okay have done the arg array in `exec_safely`. I have extracted the path out into a variable at the beginning of this function scope, I dont think it needs to be abstracted out any further though since this is just the root folder for nginx, `docker cp` just doesnt seem naturally replicated in `docker-py` sadly, it is a shame!